funsor/domains.py

```diff
@@ -2,79 +2,149 @@
 # SPDX-License-Identifier: Apache-2.0
 
 import operator
-from collections import namedtuple
+import warnings
 from functools import reduce
+from weakref import WeakValueDictionary
 
 import funsor
 import funsor.ops as ops
-from funsor.util import broadcast_shape, get_tracing_state, lazy_property, quote
+from funsor.util import broadcast_shape, get_tracing_state, quote
 
 
-class Domain(namedtuple('Domain', ['shape', 'dtype'])):
-    """
-    An object representing the type and shape of a :class:`Funsor` input or
-    output.
-    """
-    def __new__(cls, shape, dtype):
-        assert isinstance(shape, tuple)
+class Domain(type):
```

Comment:
Would it be more Pythonic to define `Domain` with `typing.Generic`?
```py
from typing import Generic, Tuple, TypeVar, Union
from typing_extensions import Literal  # backport of typing.Literal in 3.8

S = TypeVar("S", bound=Tuple)
D = TypeVar("D", bound=Union[int, str])

class Domain(Generic[S, D]):
    def __init__(self, shape: S, dtype: D):
        self.shape = shape
        self.dtype = dtype

# version 1, more concise
Bint = Domain[Tuple[()], TypeVar("DB", bound=int)]
Reals = Domain[S, Literal['real']]

# version 2, nicer printing/pickling
class Bint(Domain[Tuple[()], TypeVar("DB", bound=int)]):
    pass

class Reals(Domain[S, Literal['real']]):
    pass

def reals(*shape):
    return Reals[Tuple[tuple(Literal[size] for size in shape)]]

def bint(size):
    return Bint[Literal[size]]
```

Reply:
My longer term vision is for `Domain` to be the type of more general ground values in Funsor including `Tuple[]`, eventually including pytrees. Therefore I think it no longer makes sense to paramterize `Domain` by (`shape`, `dtype`).

Indeed I feel we are already shoe-horning `bint()` and `reals()` into an ill-fitting abstraction. For example to address #322 we will probably need heterogeneous `size` attribute which is currently aliased to `dtype`. Even that alias feels like a hack to me.

Reply:
> My longer term vision is for Domain to be the type of more general ground values in Funsor including Tuple[], eventually including pytrees

I'm definitely on board with this, but it seems like in that vision there's still a need for something like `Reals` and `Bint` (and hence my version of `Domain` above, which we could call `Shape` or `Array` after Numpy?) to represent tensor output types.  We could just have the more general ground value base type be `typing.Any`, or a very broad `TypeVar` like `TypeVar("A", bound=Union[int, str, Shape, Tuple])`?

Reply:
Hmm, I guess there's is a future where `Bint` and `Real` are more similar... I'll have to think about this more. I really don't like the way we currently use `domain.dtype` as an alias for `size`, and I'd prefer to pull these concepts apart.

Reply:
I'm fine with separating `Bint` and `Real`, it's the `BintType`/`RealType` metaclasses in this implementation that I am wondering if we could replace with something more Pythonic based on `typing.Generic` and `typing.Literal`. To clarify, this version of `Reals` simplified from my previous comment inherits directly from `Generic`:
```py
S = TypeVar("S", bound=Tuple[int, ...])

class Reals(Generic[S]):  # subtypes are cons-hashed
    def __init__(self, shape: S):
        self.shape = shape
        # dtype, etc. here

def reals(*shape):  # for illustration, calling reals() in an annotation not compatible with static type checkers?
    return Reals[Tuple[tuple(Literal[size] for size in shape)]]
```
It's more verbose, but conforms more closely to the spirit of `typing`.  Also, I'm not sure pointwise functions annotated with the metaclass-based design would be compatible with static type checkers when checked without the `function` wrapper (no types parametrized directly by values are allowed, regardless of the details of their implementations) - to what extent should that serve as a design constraint?

Reply:
I do like the idea of being more pythonic and trying to use `Generic`. However it remains to be seen whether `Generic` can serve our purpose. Issues include: we want to memoize via `WeakValueDict` rather than `lru_cache` imho; `Generic`'s implementation changes across Python versions, so it may be difficult to override `__class_getitem__` to cast parameters to int while jitting). But you have more experience with typing than I, so let's pair code on it.

> should [compatibility with type checkers] serve as a design constraint?

Yeah good question. I guess the objective of this refactoring is to replace Funsor's bespoke type annotations with more standard annotations and thereby make it easier for new Funsor users to learn and contribute. While that suggests mypy support is a secondary goal, I think it would be appreciated by IDE-based users. Let's see how hard it is, maybe we can add a `mypy` test stage.

Also I'd like to see whether we can roughly follow [NumPy types](https://numpy.org/devdocs/reference/typing.html). Let's discuss tomorrow.

Reply:
Some other challenges with more pythonic type annotations are:
- some of `typing`'s annotations simply disappear during runtime
- there is not yet a standard way to inspect type hints; we may need to roll our own version-agnostic module similar to [typing_inspect](https://github.com/ilevkivskyi/typing_inspect)